core/services/orm/tree_traversal/RelationNode.php

```diff
@@ -187,11 +190,22 @@ protected function discover()
      */
     protected function whereQueryParams()
     {
-        return [
+        $where_params =  [
             $this->related_model->get_foreign_key_to(
                 $this->main_model_obj->get_model()->get_this_model_name()
             )->get_name() => $this->main_model_obj->ID()
         ];
+        try {
+            $relation_settings = $this->main_model_obj->get_model()->related_settings_for($this->related_model->get_this_model_name());
+        } catch (EE_Error $e) {
+            // This will happen for has-and-belongs-to-many relations, when this node's related model is that join table
+            // which hasn't been explicitly declared in the main model object's model's relations.
+            $relation_settings = null;
+        }
+        if ($relation_settings instanceof EE_Has_Many_Any_Relation) {
+            $where_params[ $this->related_model->get_field_containing_related_model_name()->get_name() ] = $this->main_model_obj->get_model()->get_this_model_name();
+        }
+        return $where_params;
     }
```

Comment:
I made this change because I realized has-many-any relations (like from event to event meta) could select the wrong things if we weren't also adding a WHERE condition to use the right model name in the model name field (eg settings `'EXM_type'='Event'`)